src/storage/mvcc/reader.rs

```diff
@@ -401,9 +407,74 @@ impl MvccReader {
                     },
                 }
             };
-            if let Some(v) = self.get(&key, ts)? {
+
+            if let Some(v) = self.reverse_get_impl(&key, ts)? {
                 return Ok(Some((key, v)));
             }
+            if !self.write_cursor.as_mut().unwrap().valid() {
+                write_valid = false;
+            }
+        }
+    }
+
+    fn reverse_get_impl(&mut self, user_key: &Key, ts: u64) -> Result<Option<Value>> {
+        assert!(self.lock_cursor.is_some());
+        assert!(self.write_cursor.is_some());
+
+        // Check lock.
+        match self.isolation_level {
+            IsolationLevel::SI => {
+                let lock = {
+                    let l_cur = self.lock_cursor.as_mut().unwrap();
+                    if l_cur.valid() && l_cur.key() == user_key.encoded().as_slice() {
+                        Some(Lock::parse(l_cur.value())?)
+                    } else {
+                        None
+                    }
+                };
+                if let Some(l) = lock {
+                    self.check_lock_impl(user_key, ts, l)?;
```

Comment:
Can't we just use `check_lock` here?

Reply:
This can eliminate a `load_lock` call since we have already seek the lock